### PR TITLE
make similar return a Nemo matrix when possible

### PR DESCRIPTION
**Pull request by rfourquet**
E.g. now `similar(matrix(QQ, 1, 1, [1]), ZZ) isa fmpz_mat`.

### CODE REVIEW

**Comment by rfourquet** [comment]
Fixes https://github.com/Nemocas/AbstractAlgebra.jl/issues/407

**Comment by thofma** [comment]
Thanks for taking a look at this.

Minor remark. I think we want also `similar(::Generic.MatElem, ::FlintIntegerRing)` to also return an `fmpz_mat`. I think this is not the case with current approach right? At the moment, whenever you have something generic, similar will forever produce something generic.

So I think we want
```
similar(similar(M, R), base_ring(M)) == similar(M)
```
I guess we just need
```
similar(M::AbstractAlgebra.Generic.MatrixElem, ::FlintIntegerRing, r, c) = fmpz_mat(r, c)
```
etc.

 But maybe this is already doing this and I misunderstood :)

**Comment by rfourquet** [comment]
> Minor remark. I think we want also similar(::Generic.MatElem, ::FlintIntegerRing) to also return an `fmpz_mat`. 

Yes I thought about that, and I'm not totally sure. I think `similar` is supposed to return a matrix "connected" to the input one. For example, if `M isa MatSpaceElem{fmpz}` (it's not easy to construct such a matrix!), then I feel that `similar(M)` should return a matrix of the same type, and not a `fmpz_mat`. On the other hand, if `M` holds a type for which no Nemo matrix exists, then it may make sense that `similar(M, fmpz)` should return a `fmpz_mat`. 

> So I think we want `similar(similar(M, R), base_ring(M)) == similar(M)`

I think this is what we have with this PR: e.g. if `M isa MatElem{fmpz}`, then `similar(M)` returns a matrix of the same type, `N = similar(M, AbtractAlgebra.QQ) isa MatElem{Rational{BigInt}}` and then `similar(N, base_ring(M))` again gives something of the same type as `M`.

Are there features available on generic matrices not available for Nemo matrices? 
What do you think about keeping it like in this PR and see if it causes problems in practice?
I will ask on discourse or slack some clarification about the rule for `similar`.

**Comment by thofma** [comment]
I can tell you already that this will cause problems. This has been the behavior so far and we are relying on it explicitly (not with the similar function of course, but with the `zero_matrix` that has no changed also).

Sometimes you have to create a similar matrix say for an `fmpz_mat` and a Nemo type, for which there is no Nemo matrix type, say `nf_elem`. Then later you want to go back via `similar(.., FlintZZ)` and suddenly you are stuck with slow and useless generic matrices :)

> > So I think we want `similar(similar(M, R), base_ring(M)) == similar(M)`
> 
> I think this is what we have with this PR: e.g. if `M isa MatElem{fmpz}`, then `similar(M)` returns a matrix of the same type, `N = similar(M, AbtractAlgebra.QQ) isa MatElem{Rational{BigInt}}` and then `similar(N, base_ring(M))` again gives something of the same type as `M`.

No, I think we don't have this here. If `M isa fmpz_mat` and `K isa AnticNumberField`, then
`similar(similar(M, K), FlintZZ)` is not of type `fmpz_mat`.

**Comment by rfourquet** [comment]
I understand the use-case yes, I guess my resistance comes from that I didn't see this behavior (for `AbstractArray`s) somewhere else, but I guess this is fine, as it's hard to come up with a case where this is not what the user would want.

About:

> So I think we want similar(similar(M, R), base_ring(M)) == similar(M)

Of course you are right :) The surprising consequence is that `similar(M)` may not be of the same type as `M` (e.g. when `M isa MatSpaceElem{fmpz}`), but not a big deal. Alright I will update.

PS: I above wrote sometimes `MatElem` when I meant `MatSpaceElem`, i.e. generic matrices... I will edit the posts.

**Comment by thofma** [comment]
O know that you mean with MatrixElem. 

Can we somehow preserve that additionally `similar(M)` returns the same type as `M`?

**Comment by rfourquet** [comment]
> Can we somehow preserve that additionally similar(M) returns the same type as M?

It would complicate things, now all `similar` methods call out to `similar(mat, ring, r, c)`, which is the only one which has to be implemented by matrices. To preserve the type with `similar(M)`, it would require to implement `similar(mat, ring, r, c)`, `similar(mat)` and `similar(mat, r, c)`. This is more work, but worse in my opinion, this becomes a corner case which causes surprise. Also this would break the `similar(similar(M, R), base_ring(M)) == similar(M)` type equality.

I think a convincing use-case is warranted to implement this corner-case.

**Comment by thofma** [comment]
You are right, it is not worth it.

Ping me, once you had time to make the adjustments :)

**Comment by rfourquet** [comment]
I used `MatElem` instead of `MatrixElem`, as otherwise there is an ambiguity. It's easily resolvable, but I wonder what is wanted here. If I'm not mistaken, Nemo doesn't have matrice algebras, so do we want `similar(matalg, fmpz)` to return a `fmpz_mat`, which would loose the "algebra" character of `matalg` ?

**Comment by wbhart** [comment]
Eventually we will want Nemo to have matrix algebras. But for the time being there are only generic ones, and I guess if you start off with a generic one, you should end with a generic one.

Usually calling AbstractAlgebra.ParentConstructorFunction will construct the right thing (instead of calling ParentConstructorFunction directly). There are some exceptions in Singular.jl, but that is another story.

File: test/arb/acb_mat-test.jl [hunk 1]
```diff
@@ -116,7 +115,9 @@ end
       @test size(t) == (2, 3)
    end
 
-   println("PASS")
+   t = similar(s, CC, 2, 3)
+   @test t isa acb_mat
+   @test size(t) == (2, 3)
```
Comment (rfourquet) [inline test/arb/acb_mat-test.jl]: These 3 lines should not have been re-added, as the same test is included in the loop `for (R, M) in ring_to_mat`.  Same in other files, I will force-push on the branch to fix the rebase. But thanks anyway!

**Comment by thofma** [comment]
Oops, sorry!

**Pull request merged**